decompy/ai-rl/RewardFunction.cpp

```diff
@@ -0,0 +1,205 @@
+//https://stackoverflow.com/questions/41760481/what-should-i-replace-getglobalcontext-with-in-llvm-3-9-1
+
+    Module* mod = new Module(name, TheContext);
+    //doesn't return a function. will return cast of existing function if it is there
+    Constant* c = mod->getOrInsertFunction(name, IntegerType::get(TheContext, 32), IntegerType::get(TheContext, 32), IntegerType::get(TheContext, 32), IntegerType::get(TheContext, 32));
+    Function* mul_add = cast<Function>(c);
+    mul_add->setCallingConv(CallingConv::C); // set the calling convention to C
+
+    Function::arg_iterator args = mul_add->arg_begin(); //returns an iterator for all of the args  in the functiion
+
+    //the following names the arguments. Not strictly necessary, LLVM will name them
+    Value* x = args++; //Sets equal to args, then increments to the next arg
+    x->setName("x"); //sets name
+    Value* y = args++; //Sets equal to args, then increments to the next arg
+    y->setName("y"); //sets name
+    Value* z = args++; //Sets equal to args, then increments to the next arg
+    z->setName("z"); //sets name
+    //now we will keep the x, y, and z, pointers because the will be used later
+
+    //std::cout <<
+
+    //BasicBlock are... The basic building blocks of a program. Every function has one (the stuff between the curly braces
+    //this function needs on, so we make one:
+    BasicBlock* block = BasicBlock::Create(TheContext, "entry", mul_add);
+    //unless a lot of control is needed, us IR Builder
+    IRBuilder<> builder(block); //convinience interface for creating instructions.
+
+    //creates a binary operation. In this case, it is a multiplication instruction
+    //the builder creates and appends this instruction to the end of the block
+    // it returns the value returned by the instruction.
+    Value* tmp = builder.CreateBinOp(Instruction::Mul, x, y, "tmp");
+    //this is a return instruction
+    builder.CreateRet(tmp);
+
+    return mul_add; // now the module is complete!
+}
+
+int main(){
+    std::cout << "making first fnc" << std::endl ;
+    Function* fnc = makeLLVMFunction1("mod1");
+    std::cout << "making second fnc" << std::endl ;
+    Function* fnc2 = makeLLVMFunction2("mod2");
+
+    std::cout << "making reward" << std::endl ;
+    Reward r = Reward();
+    std::cout << "run instruction count reward" << std::endl;
+    int icr = r.instructionCountReward(*fnc, *fnc2);
+    std::cout << "The percent difference in instruction count should be 33%, the reward should be -33: " << icr <<
+    std::endl;
+
+    std::cout << "run identical instruction count reward" << std::endl;
+    int iicr = r.identicalInstructionTypeCountReward(*fnc, *fnc2);
+    std::cout << "The number of identical instruction types should be 2, the reward is 200: " << iicr << std::endl;
+
+    int totalReward = r.getReward(*fnc, *fnc2);
+    std::cout << "The total reward is 167: " << totalReward << std::endl;
+
+    //for some reason I can't delete both fncs
```

Comment:
Not deleting both functions is odd. Could this be a problem in the future?

Reply:
Will fix when we do the actual implementation